lib/que/poller.rb

```diff
@@ -265,5 +263,21 @@ def cleanup(connection)
         SQL
       end
     end
+
+    private
+
+    def any_priority_satisfied?(priorities, jobs)
+      job_priorities = jobs.map { |job| job.fetch(:priority) }.sort
+      worker_job_counts = Hash.new(0)
+      priorities.any? do |worker_priority, waiting_workers_count|
+        waiting_workers_count.times do
+          return false if job_priorities.empty?
+          break if job_priorities.first > worker_priority
+          job_priorities.shift
+          worker_job_counts[worker_priority] += 1
+        end
+        worker_job_counts[worker_priority] == waiting_workers_count
+      end
```

Comment:
Yeah, I see how this is more complicated than it first sounded, nice work.

I want to understand the relationship between the buffer size, the `available_priorities` hash and what is returned by the poller a bit better. I'm wondering whether a better way to determine whether a poll is satisfied is if we collect at least enough jobs to fill whatever buffer space is available (and maybe at least one priority?, but that feels like an anti-optimisation given the complexity of implementing it). My understanding at the moment is that the counts in `job_buffer.available_priorities` sum to the number of available workers plus the available space in the buffer. If we can fill the buffer then I think that represents a satisfactory poll, regardless of whether some workers didn't get a job.

Reply:
Just had another thought, maybe all that we care about is that the lowest priority worker queue is full?

The available space in the buffer is added to the [lowest priority worker queue](https://github.com/que-rb/que/blob/master/lib/que/job_buffer.rb#L124) when constructing the `available_priorities` hash. So if the lowest priority worker queue is full then the buffer will be full and all available workers of the lowest priority will have a job to do. I think that is a pretty good metric to use for `last_poll_satisfied`.

Reply:
Just thinking this out:

#### Five available p1 workers, five available p50 workers, 100 p1 jobs, 3 p2 jobs, buffer size of 10
{ p1: 5, p50: 15 }, so lowest priority queue is satisfied

#### Five available p1 workers, five available p50 workers, 3 p1 jobs, 100 p2 jobs, buffer size of 10
{ p1: 3, p50: 15 }, so lowest priority queue is satisfied

#### Five available p1 workers, five available p50 workers, 10 p1 jobs, 10 p2 jobs, buffer size of 10
{ p1: 5, p50: 15 }, so lowest priority queue is satisfied

#### Five available p1 workers, five available p50 workers, 9 p1 jobs, 10 p2 jobs, buffer size of 10
{ p1: 5, p50: 14 }, so lowest priority queue is *NOT* satisfied, but I think I am OK with that.

Reply:
But, maybe that doesn't actually simplify anything because to determine whether the lowest priority has been satisfied we need to do exactly what you have done and allocate jobs to higher priorities.

Another thought: a more fuzzy way of doing this is to just compare the number of jobs to the lowest priority availability without being too exact about where the jobs will actually end up?

Just thinking this out:

Five available p1 workers, five available p50 workers, available buffer space of 10, the available priorities will be
`{ p1: 5, p50: 15 }`

| p1 jobs| p2 jobs | total job | total jobs > p50 availability? (i.e. last_poll_satisfied?) |
| --- | --- | --- | --- |
| 100 | 3 | 20 | yes |
| 3   | 100 | 18 | yes |
| 10 | 10 | 15 | yes |
| 9 | 10 | 19 | yes |
| 4 | 10 | 14 | no |
| 10 | 4 | 14 | no |

That looks pretty good to me. Will see if I can break it.

Reply:
See #349

Reply:
I wanted a simpler way to understand the satisfaction determination, so tried to come up with a way to visualise it; and I think [this](https://docs.google.com/drawings/d/1F6J9MB_uDMrZ0RAVxXLRkwao8Bc8pLrVEx7lbEKOGko/edit?usp=sharing) is pretty good:

![Que poller_ worker priorities satisfaction (1)](https://user-images.githubusercontent.com/2916331/156685586-61fd4461-f9bd-4810-8aa0-35f5b8799b1b.svg)

The requested priorities are represented as a physical container, with sections of increasing width. The width of each section corresponds to the priority (priority 1 is thin, 2 is wider, 3 would be even wider, etc.). The height of each section corresponds to the number of that priority requested.

The jobs are represented as physical blocks, of varying width. The width of a block corresponds to the job's priority (priority 1 is thin, 2 is wider, 3 would be even wider, etc.).

Allocation happens by ordering the blocks by width and then dropping them into the container against its left edge. Blocks fall down until they can't squeeze through any further. The allocations are then which blocks sit in which section.

Determining the satisfaction (whether any requested priority is full of jobs) is done by seeing if any section is vertically full.

Reply:
I don't think your simplified algorithm in https://github.com/que-rb/que/pull/349 meets the satisfaction requirements. Consider the first red case in my diagram. https://github.com/que-rb/que/pull/349 would mistakenly consider that as satisfied, since the jobs count (3) is equal to the requested count of the lowest priority (p2: 3).

Reply:
Yeah, the algorithm in #349 is fuzzy, but I think it offers a very simple good-enough algorithm at the cost of an occasional re-poll when there may not be a huge number of jobs to work. 

I have tried pretty hard to think of a case where the fuzziness actually results in a problematic behaviour (i.e. a performance hit), and I haven't really come up with anything. Worst case scenario is when there are a lot of high priority workers, few lower priority workers, no/small buffer and a steady stream of high priority jobs that is not quite enough to saturate the high priority workers but more than enough to saturate the low priority workers. In that scenario we would continually be polling and only actually seeing some of the high priority workers actually do anything. It's a fine balance though, as soon as the stream of jobs becomes more than enough for the high priority workers to handle or slow enough for the number of available jobs each poll to drop below the number of low priority workers plus the buffer it will behave as expected.

Reply:
If the number of low priority workers plus the buffer size is greater than the total number of workers (as is the case with the default settings `priorities: 10,20,50,any,any,any and buffer: 8`) then it's very difficult to think of a case where you get more than one potentially unnecessary poll occasionally.

Reply:
Just to clarify, #349 is never intended to answer the question of `whether any requested priority is full of jobs`. It's a different question, which I think provides a close enough approximation.

Reply:
My main concerns with this algorithm are the ones you raised yourself: that it's complex, a bit impenetrable and that we're re-implementing worker allocation which introduces a coupling we'd probably rather not have. Happy to go with it though if you don't think #349 is appropriate.

Reply:
Ah, I see. I suppose for a rough algorithm, it is better that it err on the side of caution - sometimes polling again unnecessarily, rather than mistakenly sleeping.

Your code is way more readable than mine. And yeah, I can't see that there's much of a drawback either. So I'm leaning towards going with that. But I'd like to let it sit for a while to see how I feel about it later.

> if there is a steady stream of jobs one could argue that polling continuously is actually the preferred behaviour, even if workers are not being fully utilised?

I would have thought the notify would handle most/all of that case; but I'm not totally across it.

Reply:
> I'd like to let it sit for a while to see how I feel about it later.

Sure, sounds good.

> I would have thought the notify would handle most/all of that case; but I'm not totally across it.

Oh yeah! I completely forgot about that, you’re probably right.